net: QEMU Ethernet drivers are flaky (seemingly after "net_buf" refactor)

Testing 3 qemu ethernet drivers we added in 1.14 timeframe with the idea to use them to test the IP stack better.

master: e731bdc81a3b2e2fdf498bccab9627b18fa38de6
sample: samples/net/sockets/dumb_http_server
test command on host: ab -n1000 http://192.0.2.1:8080/


1. BOARD=qemu_x86 OVERLAY=overlay-e1000.conf
Not a single request succeeds:
~~~
$ ab -n1000 http://192.0.2.1:8080/
...
Benchmarking 192.0.2.1 (be patient)
apr_socket_recv: Connection reset by peer (104)
~~~

On Zephyr console (after ~5 `ab` attempts):

~~~
Booting from ROM..***** Booting Zephyr OS zephyr-v1.13.0-5221-g8ecd348022f9 *****
[00:00:00.019,243] <inf> net_config: Initializing network
[00:00:00.021,990] <inf> net_config: IPv4 address: 192.0.2.1
Single-threaded dumb HTTP server waits for a connection on port 8080...
[00:00:59.445,932] <err> eth_e1000: Out of memory for received frame
~~~

2. BOARD=mps2_an385 OVERLAY=overlay-smsc911x.conf
Zephyr:
~~~
Connection #137 from 192.0.2.2
Connection from 192.0.2.2 closed
Connection #138 from 192.0.2.2
Connection from 192.0.2.2 closed
Connection #139 from 192.0.2.2
Connection from 192.0.2.2 closed
Connection #140 from 192.0.2.2
...(lockup, pings don't work)...
~~~

3. BOARD=qemu_cortex_m3 OVERLAY=overlay-qemu_cortex_m3_eth.conf
Build error:
~~~
[ 54%] Building C object zephyr/CMakeFiles/zephyr.dir/drivers/ethernet/eth_stellaris.c.obj
/home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/drivers/ethernet/eth_stellaris.c: In function 'eth_stellaris_stats':
/home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/drivers/ethernet/eth_stellaris.c:291:24: error: 'struct eth_stellaris_runtime' has no member named 'stats'
  return &(DEV_DATA(dev)->stats);
                        ^~
/home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/drivers/ethernet/eth_stellaris.c: At top level:
/home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/drivers/ethernet/eth_stellaris.c:349:2: error: unknown field 'get_stats' specified in initializer
  .get_stats = eth_stellaris_stats,
  ^
/home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/drivers/ethernet/eth_stellaris.c:349:15: warning: initialization from incompatible pointer type [-Wincompatible-pointer-types]
  .get_stats = eth_stellaris_stats,
               ^~~~~~~~~~~~~~~~~~~
/home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/drivers/ethernet/eth_stellaris.c:349:15: note: (near initialization for 'eth_stellaris_apis.start')
/home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/drivers/ethernet/eth_stellaris.c: In function 'eth_stellaris_stats':
/home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/drivers/ethernet/eth_stellaris.c:292:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
~~~

--- a/subsys/net/ip/tcp.c
+++ b/subsys/net/ip/tcp.c
@@ -2338,7 +2338,7 @@ NET_CONN_CB(tcp_syn_rcvd)
 	/*
 	 * If we receive ACK, we go to ESTABLISHED state.
 	 */
-	if (NET_TCP_FLAGS(tcp_hdr) == NET_TCP_ACK) {
+	if (NET_TCP_FLAGS(tcp_hdr) & NET_TCP_ACK) {
 		struct net_context *new_context;
 		socklen_t addrlen;
 		int ret;

